scipy/stats/_binomtest.py

```diff
@@ -313,16 +312,69 @@ def binomtest(k, n, p=0.5, alternative='two-sided'):
             # special case as shortcut, would also be handled by `else` below
             pval = 1.
         elif k < p * n:
-            i = np.arange(np.ceil(p * n), n+1)
-            y = np.sum(binom.pmf(i, n, p) <= d*rerr, axis=0)
+            y = n-_binary_search_for_binom_tst(lambda x1: binom.pmf(x1, n, p),
+                                               d*rerr, np.ceil(p * n), n)+1
             pval = binom.cdf(k, n, p) + binom.sf(n - y, n, p)
         else:
-            i = np.arange(np.floor(p*n) + 1)
-            y = np.sum(binom.pmf(i, n, p) <= d*rerr, axis=0)
+            y = _binary_search_for_binom_tst(lambda x1: binom.pmf(x1, n, p),
+                                             d*rerr, 0, np.floor(p * n) + 1,
+                                             True)+1
             pval = binom.cdf(y-1, n, p) + binom.sf(k-1, n, p)
 
         pval = min(1.0, pval)
 
     result = BinomTestResult(k=k, n=n, alternative=alternative,
                              proportion_estimate=k/n, pvalue=pval)
     return result
+
+
+def _binary_search_for_binom_tst(a, d, lo, hi, asc_order=False):
+    r"""
+    Conducts an implicit binary search on a funcion specified by `a.
+
+    Meant to be used on the binomial PMF for the case of two-sided tests
+    to obtain the value on the other side of the mode where the tail prbability
+    should be computed. Because the values past the mode are always in
+    ascending or descending order, this makes binary search applicable.
+
+    Parameters
+    ----------
+    a : function
+      The function over which to perform binary search. Its values
+      for inputs lo and hi should be in ascending or descending
+      order as required by `asc_order.
```

Comment:
I'm confused by the requirement here, is this that `a(lo) <= a(mid) <= a(hi)` ***for all** values of `lo, mid, hi` in the domain of the binomial distribution or something else?

Reply:
The requirement is that a(lo) <= a(x) <= a(hi) for all lo<=x<=hi or a(lo)=>a(x)=>a(hi), depending on the boolean switch. Just like with binary search, you need a sorted array between the inputs you pass to it.